lineapy/graph_reader/artifact_collection.py

```diff
@@ -0,0 +1,335 @@
+                # art = get(**args)
+                version = args.get("version", None)
+                if version is None:
+                    art = get(artifact_name=args["artifact_name"])
+                else:
+                    art = get(
+                        artifact_name=args["artifact_name"],
+                        version=int(version),
+                    )
+                if args["artifact_name"] in self.art_name_to_node_id.keys():
+                    logger.error("%s is duplicated", args["artifact_name"])
+                    raise KeyError("%s is duplicated", args["artifact_name"])
+                self.art_name_to_node_id[args["artifact_name"]] = art._node_id
+                self.node_id_to_session_id[art._node_id] = art._session_id
+            except Exception as e:
+                logger.error("Cannot retrive artifact %s", art_entry)
+                raise Exception(e)
+
+            # Put artifact in the right session group
+            artifacts_by_session[art._session_id] = artifacts_by_session.get(
+                art._session_id, []
+            ) + [art]
+
+        # For each session, construct SessionArtifacts object
+        for session_id, session_artifacts in artifacts_by_session.items():
+            self.session_artifacts[session_id] = SessionArtifacts(
+                session_artifacts
+            )
+
+    def _sort_session_artifacts(
+        self, dependencies: TaskGraphEdge = {}
+    ) -> List[SessionArtifacts]:
+        """
+        Use the user-provided artifact dependencies to
+        topologically sort a list of SessionArtifacts objects.
+        Raise an exception if the graph contains a cycle.
+        """
+        # Construct a combined graph across multiple sessions
+        combined_graph = nx.DiGraph()
+        for session_artifacts in self.session_artifacts.values():
+            session_graph = session_artifacts.graph.nx_graph
+            combined_graph.add_nodes_from(session_graph.nodes)
+            combined_graph.add_edges_from(session_graph.edges)
+
+        # Augment the graph with user-specified edges
+        dependency_edges = list(
+            chain.from_iterable(
+                (
+                    (
+                        self.art_name_to_node_id.get(artname, None),
+                        self.art_name_to_node_id.get(to_artname, None),
+                    )
+                    for artname in from_artname
+                )
+                for to_artname, from_artname in dependencies.items()
+            )
+        )
+        if None in list(chain.from_iterable(dependency_edges)):
+            raise KeyError(
+                "Dependency graph includes artifacts not in this artifact collection."
+            )
+        combined_graph.add_edges_from(dependency_edges)
+
+        # Check if the graph is acyclic
+        if nx.is_directed_acyclic_graph(combined_graph) is False:
+            raise Exception("Provided dependencies result in a cyclic graph.")
```

Comment:
Question about the wording of this error here. 

The users define the list of dependencies they want right? If theres a cycle what this actually means is that the way they specified their dependencies conflicts with how the dependencies actually are in the LineaGraph

Wouldn't it be more clear if the error was more like "Linea detected conflict with provided dependencies"? Feel like that gets at the issue more clearly than saying theres a cycle in the graph ...

Reply:
Great suggestion - I agree that the proposed phrasing is more clear & instructive for the user. Will change the text.